### PR TITLE
Additional fleshing out of Trill provider API

### PR DESCRIPTION
**Pull request by cybertyche**
This group of refactoring fleshes out the available API. Many of the remaining methods can simply be macros on top of existing methods.

Remaining to do are to figure out what to do with Multicast and where to put certain parts of code.

### CODE REVIEW

File: Sources/Core/Microsoft.StreamProcessing.Provider/Extensions/Linq.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+using System.Linq.Expressions;
+
+namespace Microsoft.StreamProcessing.Provider
+{
+    /// <summary>
+    ///
+    /// </summary>
+    public struct Empty
+    { }
+
```
Comment (peterfreiling2) [inline Sources/Core/Microsoft.StreamProcessing.Provider/Extensions/Linq.cs]: Can we make this private?

File: Sources/Core/Microsoft.StreamProcessing.Provider/Extensions/Linq.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        /// <typeparam name="TResult"></typeparam>
+        /// <param name="left"></param>
+        /// <param name="right"></param>
+        /// <param name="resultSelector"></param>
+        /// <returns></returns>
+        public static IQStreamable<TResult> SelectMany<TLeft, TRight, TResult>(
+            this IQStreamable<TLeft> left,
+            Func<Empty, IQStreamable<TRight>> right,
+            Expression<Func<TLeft, TRight, TResult>> resultSelector)
+            => Join(left, right(default), o => true, o => true, resultSelector);
```
Comment (peterfreiling2) [inline Sources/Core/Microsoft.StreamProcessing.Provider/Extensions/Linq.cs]: Why not just have "right" be a Func<IQStreamable<TRight>> or even just IQStreamable<TRight>?

File: Sources/Core/Microsoft.StreamProcessing.Provider/Extensions/Linq.cs (hunk 1 above)
Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing.Provider/Extensions/Linq.cs]: It's temporary. We should decide whether the Empty type lives in the engine or the provider DLL and put it in only one place. Alternatively, we may not need it at all in the provider DLL if we do something different with the SelectMany API.

File: Sources/Core/Microsoft.StreamProcessing.Provider/Extensions/Linq.cs (hunk 2 above)
Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing.Provider/Extensions/Linq.cs]: This is a requirement of the LINQ syntactic sugar.

Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing.Provider/Extensions/Linq.cs]: To be more specific:

var foo = from s1 in streamable1 from s2 in streamable2...

compiles down to this and requires a func with a parameter.

File: Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs [hunk 3]
```diff
@@ -12,11 +13,49 @@ namespace Microsoft.StreamProcessing.Provider
     /// </summary>
     public abstract class QStreamableVisitor : ExpressionVisitor
     {
-        private static readonly MethodInfo AlterDurationMethod = typeof(QStreamableStatic).GetMethod("AlterDuration");
-        private static readonly MethodInfo GroupByMethod = typeof(QStreamableStatic).GetMethod("GroupBy");
-        private static readonly MethodInfo QuantizeLifetimeMethod = typeof(QStreamableStatic).GetMethod("QuantizeLifetime");
-        private static readonly MethodInfo SelectMethod = typeof(QStreamableStatic).GetMethod("Select");
-        private static readonly MethodInfo WhereMethod = typeof(QStreamableStatic).GetMethod("Where");
+        private static readonly Expression<Func<IQStreamable<object>, IQStreamable<object>>> ChopExpression = s => s.Chop(0, 0);
+        private static readonly MethodInfo ChopMethod = ((MethodCallExpression)ChopExpression.Body).Method.GetGenericMethodDefinition();
```
Comment (peterfreiling2) [inline Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: Are the "Expression<Func<...>>" members needed only to construct the MethodInfo members? If so, maybe a static constructor so we don't have to keep the former around as a member?

File: Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs [hunk 4]
```diff
@@ -12,11 +13,49 @@ namespace Microsoft.StreamProcessing.Provider
     /// </summary>
     public abstract class QStreamableVisitor : ExpressionVisitor
     {
-        private static readonly MethodInfo AlterDurationMethod = typeof(QStreamableStatic).GetMethod("AlterDuration");
```
Comment (peterfreiling2) [inline Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: >AlterDurationMethod [](start = 43, length = 19)

What happened to AlterDuration?

File: Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs (hunk 3 above)
Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: I could go either way on that, though StyleCop at one point frowned on a static constructor for some reason.

Comment (peterfreiling2) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: Interesting, I wonder why

---
In reply to: [309311070](https://github.com/microsoft/Trill/pull/103#discussion_r309311070) [](ancestors = 309311070)

File: Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs (hunk 4 above)
Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: It's still there somewhere - was it renamed?

File: Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs (hunk 3 above)
Comment (peterfreiling2) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: Was it SA1642?
If the constructor is static, the summary text should begin with "Initializes static members of the {class name} class." For example:

/// <summary>
/// Initializes static members of the <see cref="Customer"/> class.
/// </summary>
public static Customer()
{
}

---
In reply to: [309311333](https://github.com/microsoft/Trill/pull/103#discussion_r309311333) [](ancestors = 309311333,309311070)

File: Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs (hunk 4 above)
Comment (peterfreiling2) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: Looks completely gone to me

---
In reply to: [309311687](https://github.com/microsoft/Trill/pull/103#discussion_r309311687) [](ancestors = 309311687)

File: Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs (hunk 3 above)
Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: No, the one I saw at the time said don't use them. That said, I think that was the old FxCop framework so I'm not sure.

File: Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs (hunk 4 above)
Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: It's called SetDuration now. If I recall correctly, I was trying to use "set" when setting things to constant values and "alter" if variable to set a consistent naming scheme.

Comment (peterfreiling2) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: Ah got it, makes sense

---
In reply to: [309313151](https://github.com/microsoft/Trill/pull/103#discussion_r309313151) [](ancestors = 309313151)

File: Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs (hunk 3 above)
Comment (peterfreiling2) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: CA1810 - learned something new

---
In reply to: [309312580](https://github.com/microsoft/Trill/pull/103#discussion_r309312580) [](ancestors = 309312580)

Comment (peterfreiling2) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: How about this?

        private static readonly MethodInfo ChopMethod = ConvertExpressionToMethodInfo(s => s.Chop(0, 0));
        private static readonly MethodInfo ClipDurationMethod = ConvertExpressionToMethodInfo(s => s.ClipDuration(0));
...
        private static MethodInfo ConvertExpressionToMethodInfo(Expression<Func<IQStreamable<object>, IQStreamable<object>>> expression)
            => ((MethodCallExpression)expression.Body).Method.GetGenericMethodDefinition();

---
In reply to: [309315170](https://github.com/microsoft/Trill/pull/103#discussion_r309315170) [](ancestors = 309315170,309312580)

Comment (cybertyche) [reply Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: I think that could work - I'll try it out in a supplement.

File: Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs [hunk 5]
```diff
@@ -12,11 +13,49 @@ namespace Microsoft.StreamProcessing.Provider
...
+        private static readonly Expression<Func<IQStreamable<object>, IQStreamable<IGroupedWindow<object, object>>>> GroupByExpression = s => s.GroupBy(o => o, o => o);
+        private static readonly MethodInfo GroupByMethod = ((MethodCallExpression)GroupByExpression.Body).Method.GetGenericMethodDefinition();
+
+        private static readonly Expression<Func<IQStreamable<object>, IQStreamable<object>>> SelectExpression = s => s.Select(o => o);
+        private static readonly MethodInfo SelectMethod = ((MethodCallExpression)SelectExpression.Body).Method.GetGenericMethodDefinition();
+
+        private static readonly Expression<Func<IQStreamable<object>, IQStreamable<object>>> SelectManyExpression = s => s.SelectMany(o => Array.Empty<object>());
+        private static readonly MethodInfo SelectManyMethod = ((MethodCallExpression)SelectManyExpression.Body).Method.GetGenericMethodDefinition();
+
+        private static readonly Expression<Func<IQStreamable<int>, IQStreamable<int>>> WhereExpression = s => s.Where(o => true);
```
Comment (peterfreiling2) [inline Sources/Core/Microsoft.StreamProcessing.Provider/QStreamableVisitor.cs]: >IQStreamable<int>, IQStreamable<int>> [](start = 48, length = 37)

<int>?

**Pull request merged**